### ISSUE TITLE
1d wind model failing with errors

### ISSUE THREAD

**Issue by Higginbottom** [sirocco-rt/sirocco#686]
I am attempting to import 1D Pluto models into python in order to perform rad-hydro calculations.
My initial attempt failed with many errors of the type: spherical: ds)in_cell s and max returning VERY_BIG in cell ... and eve more translate_in_wind: Grid cell not in wind - we get a million of these, and the code crashes.

Following KSLs suggestions - I have simplified the situation for testing:

1: I generate a 1D model using the following parameter file:

[mod.pf.txt](https://github.com/agnwinds/python/files/4403689/mod.pf.txt)

2: This runs fine, and produces a wind save file. This is then processed through windsave2table.

3: We then process this through import_1d.py mod which makes a model file:

[mod.import.txt](https://github.com/agnwinds/python/files/4403704/mod.import.txt)

This has one inner guard cell (with zero density) and two outer guard cells - I assume this is correct..

4: We then read *that* model into python and run it via this parameter file. It crashes.

[import.pf.txt](https://github.com/agnwinds/python/files/4403722/import.pf.txt)

concerningly, the runtime errors look like:

ds_in_cell s and smax returning VERY_BIG in cell 1397164432 nudging photon by DFUDGE

which seems to have a dodgy plasma cell number....

Anyway - confirmation from others that their installation does the same would be appreciated. Plus any input from Ed..

**Comment by Higginbottom** [comment]
Possibly related to #608

**Comment by saultyevil** [comment]
Yep, I have the same problem as you. I've got the same problem in the latest commit of dev ([508233c
](https://github.com/agnwinds/python/commit/508233cc934c40f3cee275a44e3e50b851e6374a)) and a version of dev before I messed around with the import code  ([fd84d4d](https://github.com/agnwinds/python/commit/fd84d4dfb1b0a834d0b4ac5c3537270bf5d4d7ae))

I think the problem is something to do with this Error,

```
Error: translate_in_wind: Grid cell 28 of photon is not in wind, moving photon 8.51e+11
```

I do actually think it's correct that cell 28 isn't in the wind. But I don't know why photons are going into that cell. If I reduce wind.radmax to 5e12, then the model runs without throwing up any errors. But that's clearly the wrong value of wind.radmax...

I think the problem is in spherical_make_grid_import(). zdom[ndom].rmax = r[ncell - 1] when it should be r[ncell - 2], as there are TWO guard cells. Changing that value seemed to get rid of all of the errors.

**Comment by saultyevil** [comment]
Actually, before I forget, I found a related "bug"...

We shouldn't be setting a value for Wind.radmax for imported models, since we use the boundary of the first guard cell as the maximum wind radius (zdom[ndom].rmax). 

We throw the Error, 

```
Error: get_wind_params: zdom[ndom].rmax 0 for wind type 10
```

when `if (zdom[ndom].rmax == 0)` in get_wind_params() because, unlike with the other models, the imported models do not have a set up parameters we read in to define the wind. The simple fix is to add a check for `if (zdom[ndom].rmax == 0 && zdom[ndom].wind_type != IMPORT)`. Then we won't need to supply Wind.radmax when there is no reason to.

I'll make the changes and test now.

**Comment by Higginbottom** [comment]
Ah cool - I'll let you do it :-)

**Comment by Higginbottom** [comment]
OK, there is still an issue here, and its to do with the inner edge.
First - a question - do we *need* an inner ghost/guard cell in a case like this. 

Second, the problem. If one makes a detached model - that is where you have an inner edge of the shell that is not co-located with the outer edge of the star, then the first cell is treated as a real cell, even if it has zero density etc. 

Here is how to reproduce it.

Run this parameter file to make a model with a detached shell..

[mod.pf.txt](https://github.com/agnwinds/python/files/4408720/mod.pf.txt)

this makes a shell that run from 9.6e12 to 9.6e13 - the star has a radius of 9.6e11.

Then, run windsave2table and import_1d.py and it makes this

[mod.import.txt](https://github.com/agnwinds/python/files/4408729/mod.import.txt)


the critical first few lines are:

 i                 r                  v      rho     t_e
-- ----------------- ------------------ -------- -------
 0   8384000000000.0  19940411.22946064      0.0     0.0
 1   9600000000000.0  19940411.22946064 1.22e-15 41400.0
 2  10990000000000.0  55437171.64502533 5.01e-16 41400.0


so there is a cell 0, with zero density and zero temperature that is inside the inner shell. This is the ghost cell.

Run this with 

[import.pf.txt](https://github.com/agnwinds/python/files/4408742/import.pf.txt)

and all hell breaks loose and it crashes big-style.

if you delete the last colum of import.pf.txt - i.e. removing the temperatures 

[mod.import_no_te.txt](https://github.com/agnwinds/python/files/4408747/mod.import_no_te.txt)

then it runs, but inspecting the output via py_wind, one sees that the first cell is not set as a ghost cell. 

# TITLE= "import.rho.dat"
# Coord_Sys SPHERICAL
r var inwind i
8.38e+12 0.00e+00   0   0 
9.60e+12 1.22e-15   0   1 
1.10e+13 5.01e-16   0   2 
1.26e+13 2.73e-16   0   3 

This works 'o.k.' with just zero density, but if the temperature is set to zero too, that causes the catastrophe.

If the first cell has radius inside the central object, then python correctly sets it to a ghost zone  - but we do need to be able to deal with detached shells.

So, this needs to be fixed, but the question is - what is the correct behaviour? DO we want to force the user to supply a ghost cell - if so python needs to be able to recognise its nature in some way.

OR - do we want python to be able to automagically set up ghost cells - at the start and maybe at the end of the model. This would probably be best from a user experience POV - otherwise the user has to have a good knowledge of the requirements for ghost cells. Certainly a code like Pluto sets up these cells itself...

Thoughts?

**Comment by saultyevil** [comment]
The problem here is unlike with the other coordinate systems we can import, for the spherical model we assume that all the cells are in-wind, so making a detached model isn't possible since it wasn't programmed to be like that.

In principle, and to in my opinion make it more consistent with the other input formats, we could require an inwind column like we do for cylindrical or polar models.

**Comment by Higginbottom** [comment]
Thats certainly a good solution.

**Comment by saultyevil** [comment]
@Higginbottom oki, I think I've fixed the issue with a detached wind.. but it's not pretty and quite hacky ATM until I figure it out better - I was having a bit of trouble trying to get the volume of the cell to be set to 0. I've created a new branch named bug686 which you can use for now if you want to!

**Comment by Higginbottom** [comment]
Crikey Ed - you are industrious! Was waiting for KSL to comment before working on this....
I've been getting around it by simly setting my innermost cell radius to less than the stars radius... But this is amazing work..

**Comment by saultyevil** [comment]
I don't think all of my changes are currently in the spirit of Python, but it *seems* I was able to run an imported model of the mod.pf.txt you posted earlier.. which was originally created from this pf,

[mod.pf.txt](https://github.com/agnwinds/python/files/4409268/mod.pf.txt)

**Comment by saultyevil** [comment]
Python may not crash now, but I don't think it's working properly as I'm not sure photons are going through the grid properly. Photons aren't scattering and the nphot table is empty..

```
 nphot
       0        0        0        0        0        0        0        0        0        0 
       0        0        0        0        0        0        0        0        0        0 
```

**Comment by Higginbottom** [comment]
Hmmmm, not having much luck here.

I'm trying to read in this file:


[pluto_model.dat.txt](https://github.com/agnwinds/python/files/4409451/pluto_model.dat.txt)

using this param

[input.pf.txt](https://github.com/agnwinds/python/files/4409443/input.pf.txt)


crashes rather nastily......

Have I set the inwind numbers correctly ?

second number - -1 for not in wind, 0 for in wind?

**Comment by saultyevil** [comment]
Yeah, you've set it up correctly. I think what's happening is that a plasma cell has density = -nan, which means some calculation in saha() = -nan and so it crashes.. I think this is gonna take a bit of fiddling

**Comment by Higginbottom** [comment]
My initial thought was that the code *thought* it was reading in a temperature - I got the same sort of crash when the code read in the first cell with T_e=0 thinking it was real not a ghost. But that didn't seem to be the case...
Lets see what KSL comments...

**Comment by saultyevil** [comment]
@Higginbottom Nick, can you try out the changes I've made in the Bug686 branch? It looks like I was able to run the pluto_model.dat and input.pf with the changes I made.

The problem turned out to be two things, I think. The first was that the inner boundary was being set to r[0] instead of r[1], so it was being set as the coordinate of the 1st guard cell. Secondly, there was a progamming mistake where I set the temperature to t.wind.init, thinking it was initialised but was actually still 0 at that point in the program flow. This meant when a temperature was provided, the code generally seemed to work (?) but otherwise the ion densities and etc were all being calculated at T = 0, creating all of the NaNs and failed sane_checks.

The new code doesn't read a inwind variable, as every cell (other than the ghost cells) is assumed in the wind, i.e. the wind is continous. As long as the first "real" cell starts at the wind radius, then we can have a detatched situation as described earlier as it seems the ghost cell can be as large as one wants.

I in general refactored the import code I wrote now to be more consistent with Python. You also shouldn't need to provide Wind.radmax (as it was being set by the grid dimensions anyway), and only should need to provide Wind.t.init if no temperature is provided with the read in model.

Here are the files I tested with,

Pluto:
[input.pf.txt](https://github.com/agnwinds/python/files/4437461/input.pf.txt)
[pluto_model.dat.txt](https://github.com/agnwinds/python/files/4437462/pluto_model.dat.txt)

Detached Spherical:
[import.pf.txt](https://github.com/agnwinds/python/files/4437473/import.pf.txt)
[mod.import.txt](https://github.com/agnwinds/python/files/4437475/mod.import.txt)

**Comment by Higginbottom** [comment]
Hmmm,
so I checkout bug686, and then simply do py84c input.pf - it crashes with 


There are 1 domains
                Wind type:  10  ndim: 203 mdim:   1 ndim2:  203
Allocated        312 bytes for each of   203 elements of             totaling        0.1 Mb
Define wind coord_type 0 for domain 0
wind2d: For domain 0 there are 203 cells of which 0 are in inwind, 0 partially in_wind, & 0 with pos. vol
wind2d: There are no wind cells in domain 0.  This seems unreasonble
--------------------------------------------------------------------------
Aborting rank 0000: exiting all processes with error 1

Error summary for thread 0: summary prior to abort
Recurrences --  Description
        7 -- getatomic_data: line input f odd (may be OK if Macro): %s
        1 -- Ignoring %d inner shell cross sections because no matching yields
        1 -- wind2d: There are no wind cells in domain %d.  This seems unreasonble

The other one requires an input for temperature, - but after giving it that it also crashes with lots of 
sane_check: 0 nan
coord_frac:sane_check dr=nan for spherical coords. 

and finally

wind2d: For domain 0 there are  20 cells of which 0 are in inwind, 0 partially in_wind, & 0 with pos. vol
wind2d: There are no wind cells in domain 0.  This seems unreasonble
--------------------------------------------------------------------------
Aborting rank 0000: exiting all processes with error 1

Error summary for thread 0: summary prior to abort
Recurrences --  Description
        7 -- getatomic_data: line input f odd (may be OK if Macro): %s
        1 -- Ignoring %d inner shell cross sections because no matching yields
       20 -- sane_check: %d %e
       20 -- coord_frac:sane_check dr=%f for spherical coords. 
        1 -- wind2d: There are no wind cells in domain %d.  This seems unreasonble

I assume I'm doing something wrong, but I cant see what - I am trying to run with exactly the files in the message above..


hmmmm

**Comment by Higginbottom** [comment]
Just to confirm - if I checkout dev, these two runs still work fine with those files...

**Comment by Higginbottom** [comment]
Ooops - forgot to pull in latest changes - run OK... 
Testing other stuff...

**Comment by Higginbottom** [comment]
OK, so experimenting I find that, when exporting my Pluto model I need to

define an extra inner cell - this can have the same radius as the innermost real cell

define two extra outer cells - these *cannot* have the same radius as the outermost real cell.

If the outermost cells have the same radius as the last real cell - then the last real cell *and* the two ghost cells all get defined as outside the wind. I guess that this is because the wind outer radius gets defined by the radius of the outer cell?

If this is the expected behaviour, then I can work with it - and its a find solution from my POV - but we need to write this up for future generations....

**Comment by saultyevil** [comment]
So it has been hardcoded in spherical_volumes() that the 1st cell and the last 2 cells have 0 volume and aren't in the wind for all spherical models.

I believe the expected input for imported models is the same as when we create a spherical grid, that being that the cell coordinates should always be increasing in size. The inner radius is defined by `zdom[ndom].rmin = imported_model[1].r`, to account for the first cell being a ghost cell, and the outer radius set by `zdom[ndom].rmax = imported_model[ncells - 2].r`, to account for the last two cells being ghost cells.

There's nothing to stop you right now from providing a grid where the ghost cells have the same radius at the real cells, but it still "works" as you found it... I guess it's sorta like undefined behaviour in C?

I think this suggests that we need an additional safety check for the grid coordinates, i.e. check that they increase in size.

**Comment by saultyevil** [comment]
Though looking at the 1d_sn grid, it confuses me how the inner ghost cell has the same velocity as the first real cell,
```
 i                  r                  v      rho   t_e
-- ------------------ ------------------ -------- -----
 0 1208000000000000.0  1100258151.526268      0.0   0.0
 1 1236000000000000.0  1100258151.526268 7.41e-14 40000
```
But the outer ghost cells have a different velocity to the last real cell.
```
27 2197000000000000.0  1951614716.074871 1.32e-15 40000
28 2246000000000000.0  1994041122.946064      0.0   0.0
29 2297000000000000.0 2050609665.4409878      0.0   0.0
```
I quickly experimented and it doesn't seem to matter if the outer ghost cells have the same velocity as the last real cell though.

More to add to the importing documentation!

**Comment by Higginbottom** [comment]
OK, so as you say, you need to have constantly increasing radius in the outer cells or you get extra ghost cells generated - as with this model/parameter file. You end up with three outer ghost cells...

[input.pf.txt](https://github.com/agnwinds/python/files/4443757/input.pf.txt)
[pluto_model.txt](https://github.com/agnwinds/python/files/4443758/pluto_model.txt)

**Comment by saultyevil** [comment]
I've expanded the documentation to hopefully include more gotcha's, making it easier for everyone to figure out how to import a grid.

I've also now added a check so Python will exit if a spherical model does not have a constantly increasing radius. So for the model Nick posted above, Python will exit. 

I haven't done this for polar or cylindrical imports. It's a little bit tricker, since it's possible to import an unstructured grid as long as the grid dimensions are consistent. We should decide if we want to allow unstructured 2D grids (like the original gird in Jane's model)... in principle, I have nothing against it and we can take a "use at your own caution" approach.
